From 3e065c3325c059c730df20aa15c2d0788f3e1132 Mon Sep 17 00:00:00 2001
From: Benedikt Niedermayr <benedikt.niedermayr@siemens.com>
Date: Wed, 22 Feb 2023 12:52:09 +0100
Subject: [PATCH 3/6] gpio-chardev: init compatibility

Pins initialized with mraa_gpio_chardev_init() should behave like
pins initialized with mraa_gpio_init(). This applies especially for
several attributes of the pin's mraa_gpio_context struct:

* advance_func: Required to execute custom functions of the pin's
                function table.
* phy_pin: Required to distinguish between board pins and onboard
           mulitplexer pins(pull_en, out_en)

Signed-off-by: Benedikt Niedermayr <benedikt.niedermayr@siemens.com>
---
 src/gpio/gpio.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/gpio/gpio.c b/src/gpio/gpio.c
index f13081f..b1c3d27 100644
--- a/src/gpio/gpio.c
+++ b/src/gpio/gpio.c
@@ -505,6 +505,15 @@ mraa_gpio_chardev_init(int pins[], int num_pins)
     /* Initialize events array. */
     dev->events = NULL;
 
+    /*
+     * Single pin init
+     * Stay compatible to mraa_gpio_init()
+     */
+    if (num_pins == 1) {
+        dev->advance_func = board->adv_func;
+        dev->pin = pins[0];
+        dev->phy_pin = dev->pin;
+    }
     return dev;
 }
 
-- 
2.25.1

